File: 437/437.v2.cpp
#include <cstdlib>
#include <unordered_map>

using namespace std;

struct TreeNode {
	int val;
	TreeNode *left;
	TreeNode *right;
 	TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 };

class Solution {
private:
	unordered_map<int, int> cache;
	int count; // Count of proper paths
	
	void visit(TreeNode* root, int total, int sum) {
		if (!root)
			return;
		
		total += root->val;
		
		// If paths ending on this node exist
		if (cache.find(total - sum) != cache.end()) {
			count += cache[total - sum];
		}
		
		cache[total] += 1;
		
		visit(root->left, total, sum);
		visit(root->right, total, sum);
		
		// Decrement the cache, because following DFS will be on a different route
		cache[total] -= 1;
	}
public:
	int pathSum(TreeNode* root, int sum) {
		// Init cache
		cache[0] = 1;

		visit(root, 0, sum);
		
		return count;
	}
};

File: 0400-0499/437/437.v2.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

var cache map[int]int
var count int

func visit(root *TreeNode, total int, sum int) {
	if root == nil {
		return
	}

	total += root.Val

	// If path ending on this node exists, add the number of paths to counter
	if v, ok := cache[total-sum]; ok {
		count += v
	}

	cache[total] += 1

	visit(root.Left, total, sum)
	visit(root.Right, total, sum)

	// Decrement the cache, because following DFS will be on a different route
	cache[total] -= 1
}

func pathSum(root *TreeNode, sum int) int {
	// Reset global vars
	cache = make(map[int]int)
	count = 0

	// Init cache
	cache[0] = 1

	visit(root, 0, sum)

	return count
}
